[libpas] Improve parsing of Probabilistic Guard Malloc activation state

zakariaridouh opened the pull request:
#### 262891bb0d263134adb58951a89f337996435804
<pre>
[libpas] Improve parsing of Probabilistic Guard Malloc activation state
<a href="https://bugs.webkit.org/show_bug.cgi?id=295649">https://bugs.webkit.org/show_bug.cgi?id=295649</a>
<a href="https://rdar.apple.com/155454591">rdar://155454591</a>

Reviewed by Dan Hecht.

Fix a variety of issues on the WebKit side related to reading the PGM activation state.
We need to add this flag into the pas root so we can subsequently read it from the dead process.

* Source/JavaScriptCore/API/PASReportCrashPrivate.cpp:
(PASPGMEnabledOnProcess): Deleted.
* Source/JavaScriptCore/API/PASReportCrashPrivate.h:
* Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.c:
(pas_probabilistic_guard_malloc_deallocate):
(pas_probabilistic_guard_malloc_initialize_pgm):
(pas_probabilistic_guard_malloc_enabled_on_process): Deleted.
* Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.h:
* Source/bmalloc/libpas/src/libpas/pas_report_crash.c:
(pas_report_crash_extract_pgm_failure):
* Source/bmalloc/libpas/src/libpas/pas_report_crash_pgm_report.h:
* Source/bmalloc/libpas/src/libpas/pas_root.c:
(pas_root_construct):
* Source/bmalloc/libpas/src/libpas/pas_root.h:

Canonical link: <a href="https://commits.webkit.org/298400@main">https://commits.webkit.org/298400@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5afeac2a88475725cf659205515b225b752d9ddd

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/115402 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/35096 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/25598 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/121491 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/65980 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/45af06b4-e788-4844-894d-ce0b54243496) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/35755 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/43667 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/87685 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42372 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/e00564ff-5ab8-4524-b65b-4e60416f3961) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/118350 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/28518 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/103591 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/68077 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/27676 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/21713 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/65148 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/107610 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/97907 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/21826 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/124656 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/113941 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/42341 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/31720 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/96467 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/42709 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/99779 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/96253 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24488 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/41481 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/19337 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/38251 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/42222 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/47777 "Built successfully") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/35/builds/139046 "Build was cancelled. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Failed to compile JSC; Compiled JSC (cancelled)") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/41725 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/35/builds/139046 "Build was cancelled. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Failed to compile JSC; Compiled JSC (cancelled)") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/45053 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/43444 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

zakariaridouh commented on Source/JavaScriptCore/API/PASReportCrashPrivate.h:
```diff
@@ -34,7 +34,7 @@
 #endif
 
-JS_EXPORT kern_return_t PASPGMEnabledOnProcess(void);
+
```
extra line, not needed.

zakariaridouh commented on Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.c:
```diff
@@ -391,14 +391,7 @@ pas_ptr_hash_map_entry* pas_probabilistic_guard_malloc_get_metadata_array(void)
-{
-    return pas_probabilistic_guard_malloc_has_been_used;
-}
+
```
extra line, not needed.

dhecht commented on Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.h:
```diff
@@ -102,6 +102,12 @@ struct pas_pgm_storage {
 };
 
+/* 
+ * Flag to indicate if PGM has enabled at all for this process,
```
```suggestion
 * Flag to indicate if PGM was ever enabled for this process,
```

dhecht reviewed (approved):
r=me, with nit

zakariaridouh committed:
[libpas] Improve parsing of Probabilistic Guard Malloc activation state
https://bugs.webkit.org/show_bug.cgi?id=295649
rdar://155454591

Reviewed by Dan Hecht.

Fix a variety of issues on the WebKit side related to reading the PGM activation state.
We need to add this flag into the pas root so we can subsequently read it from the dead process.

* Source/JavaScriptCore/API/PASReportCrashPrivate.cpp:
(PASPGMEnabledOnProcess): Deleted.
* Source/JavaScriptCore/API/PASReportCrashPrivate.h:
* Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.c:
(pas_probabilistic_guard_malloc_deallocate):
(pas_probabilistic_guard_malloc_initialize_pgm):
(pas_probabilistic_guard_malloc_enabled_on_process): Deleted.
* Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.h:
* Source/bmalloc/libpas/src/libpas/pas_report_crash.c:
(pas_report_crash_extract_pgm_failure):
* Source/bmalloc/libpas/src/libpas/pas_report_crash_pgm_report.h:
* Source/bmalloc/libpas/src/libpas/pas_root.c:
(pas_root_construct):
* Source/bmalloc/libpas/src/libpas/pas_root.h:

Canonical link: https://commits.webkit.org/298400@main
